pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,332 @@
+	secretObj := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretNameInGardenCluster,
+			Namespace: b.Shoot.Info.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(b.Shoot.Info, gardenv1beta1.SchemeGroupVersion.WithKind("Shoot")),
+			},
+		},
+		Type: corev1.SecretTypeOpaque,
+		Data: map[string][]byte{
+			common.EtcdEncryptionSecretFileName: ecYamlBytes,
+		},
+	}
+	if _, err := client.CreateSecretObject(secretObj, true); err != nil {
+		return err
+	}
+	b.Logger.Info("EncryptionConfiguration written to garden successfully")
+	return nil
+}
+
+// isEncryptionConfigurationConsistent checks whether the configuration is consistent in seed and garden
+func (b *Botanist) isEncryptionConfigurationConsistent(ecSeed *apiserverconfigv1.EncryptionConfiguration, ecGarden *apiserverconfigv1.EncryptionConfiguration) (bool, error) {
+	equal, err := encryptionconfiguration.Equals(ecSeed, ecGarden)
+	if (err != nil) || !equal {
+		return false, fmt.Errorf("EncryptionConfiguration in seed cluster and garden cluster are not equal: %v", err)
+	}
+	consistent, err := encryptionconfiguration.IsConsistent(ecSeed)
+	if (err != nil) || !consistent {
+		return false, fmt.Errorf("EncryptionConfiguration in seed cluster is not consistent: %v", err)
+	}
+	return true, nil
+}
+
+// needToRewriteShootSecrets checks whether the secrets in the shoot need to
+// be rewritten, e.g. after a change to the EncryptionConfiguration
+func (b *Botanist) needToRewriteShootSecrets() (needToRewriteSecrets bool, err error) {
+	client := b.Operation.K8sSeedClient
+	ecs, err := client.GetSecret(b.Operation.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, fmt.Errorf("no EncryptionConfiguration found in seed cluster (%v)", b.Operation.Shoot.SeedNamespace)
+		} else {
+			return false, err
+		}
+	}
+	if ecs.Annotations == nil {
+		// no annotations found, rewrite secrets
+		return true, nil
+	}
+	lastSecretRewriteChecksum, ok := ecs.Annotations[common.EtcdEncryptionChecksumAnnotationName]
+	if !ok {
+		// no annotation found, rewrite secrets
+		return true, nil
+	}
+	if b.CheckSums[common.EtcdEncryptionSecretName] == lastSecretRewriteChecksum {
+		// no change in checksum, no need to rewrite secrets
+		return false, nil
+	} else {
+		// checksum changed, do rewrite
+		return true, nil
+	}
+}
+
+// rewriteShootSecrets rewrites all secrets of the shoot.
+// This will take into account the current EncryptionConfiguration.
+func (b *Botanist) rewriteShootSecrets() error {
+	client := b.Operation.K8sShootClient
+	secretList, err := client.ListSecrets(metav1.NamespaceAll, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, secret := range secretList.Items {
+		secretName := secret.GetName()
+		secretNamespace := secret.GetNamespace()
+		attempt := 0
+		err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			attempt++
```

Comment:
Why is the attempt interesting? Can we remove it?

Reply:
well, I did not trust the implementation of this retry, and thus I wanted to see it in action

Reply:
Okay, I think we don't need it here, doesn't help too much I guess. Let's remove it.